crates/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -0,0 +1,328 @@
+    /// and registered as a new generation using [Self::add_generation].
+    pub fn realize_current_generation(
+        &self,
+    ) -> Result<PathEnvironment<Original>, GenerationsError> {
+        let metadata = self.metadata()?;
+        let current_gen = metadata
+            .current_gen
+            .ok_or(GenerationsError::NoGenerations)?;
+        self.realize_generation(*current_gen)
+    }
+
+    /// Import an existing environment into a generation
+    ///
+    /// Assumes the invariant that the [PathEnvironment] instance is valid.
+    ///
+    /// This will copy the manifest and lockfile from the source environment
+    /// into a generation folder.
+    /// Any other assets such as hook scripts are ignored.
+    ///
+    /// If the generation already exists, it will be overwritten.
+    ///
+    /// If `set_current` is true, the generation will also be set as the current generation.
+    fn register_generation(
+        &mut self,
+        environment: PathEnvironment<Original>,
+        generation: usize,
+        generation_metadata: GenerationMetadata,
+        set_current: bool,
+    ) -> Result<(), GenerationsError> {
+        let realized = self.realize()?;
+
+        let description = generation_metadata.description.clone();
+
+        let mut metadata = self.metadata()?;
+        let _existing = metadata
+            .generations
+            .insert(generation.into(), generation_metadata);
+
+        if set_current {
+            metadata.current_gen = Some(generation.into());
+        }
+
+        write_metadata_file(metadata, realized.path())?;
+
+        let generation_path = realized.path().join(generation.to_string());
+        let env_path = generation_path.join(ENV_DIR_NAME);
+        fs::create_dir_all(&env_path).unwrap();
+
+        // todo: we need to figure out what comes with a path env.
+        //       i.e. which files other than the manifest and lockfile do we need to copy?
+        //       what about assets such hook scripts?
+
+        // copy manifest and lockfile (if it exists)
+        // silently overrides existing files (but does not remove them if they are not in the source)
+        fs::copy(
+            environment.manifest_path(),
+            env_path.join(MANIFEST_FILENAME),
+        )
+        .unwrap();
+        if environment.lockfile_path().exists() {
+            fs::copy(
+                environment.lockfile_path(),
+                env_path.join(LOCKFILE_FILENAME),
+            )
+            .unwrap();
+        }
+
+        realized.add(&[Path::new(".")]).unwrap();
+        realized
+            .commit(&format!(
+                "Register generation {}\n\n{}",
+                generation, description
+            ))
+            .unwrap();
+        realized.push("origin", false).unwrap();
+
+        Ok(())
+    }
+
+    /// Create a new generation from an existing environment
+    ///
+    /// This will copy the manifest and lockfile from the source environment
```

Comment:
Question: Should this also mention the invariant that `environment` must be a valid `PathEnvironment`?

Reply:
adding